release/src/router/wol/doc/wol.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "WOL 1"
.TH WOL 1 "2004-04-29" "wol 0.7.1" "wol"
.SH "NAME"
wol \- Wake On LAN client
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
wol [\fI\s-1OPTION\s0\fR] ... \fIMAC-ADDRESS\fR ...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This manual gives you an introduction into \f(CW\*(C`wol\*(C'\fR, the Wake On \s-1LAN\s0 client.
It remotely turns on computers that supports Magic Packet technology
(also known as Wake On \s-1LAN\s0).
.PP
Some workstations support SecureON password feature. These machines only
wake up if you provide the correct password. \f(CW\*(C`wol\*(C'\fR also provides this
feature.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
Print a summary of the command line options.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of \f(CW\*(C`wol\*(C'\fR.
.IP "\fB\-h \s-1HOST\s0\fR" 4
.IX Item "-h HOST"
.PD 0
.IP "\fB\-\-host=HOST\fR" 4
.IX Item "--host=HOST"
.IP "\fB\-i \s-1HOST\s0\fR" 4
.IX Item "-i HOST"
.IP "\fB\-\-ipaddr=HOST\fR" 4
.IX Item "--ipaddr=HOST"
.PD
Broadcast packet to this \s-1IP\s0 address or hostname. This is important if your wol
client is a multihomed host and you want to send only to one subnet
(default \s-1IP\s0 address is \f(CW255.255.255.255\fR).
.IP "\fB\-p \s-1NUM\s0\fR" 4
.IX Item "-p NUM"
.PD 0
.IP "\fB\-\-port=NUM\fR" 4
.IX Item "--port=NUM"
.PD
Send packet with this destination port \fI\s-1NUM\s0\fR. This option is important if
your packet filter would block the default destination port \f(CW40000\fR.
.IP "\fB\-f \s-1FILE\s0\fR" 4
.IX Item "-f FILE"
.PD 0
.IP "\fB\-\-file=FILE\fR" 4
.IX Item "--file=FILE"
.PD
Read hardware addresses, \s-1IP\s0 addresses/hostnames, optional ports and SecureON
password from file \fI\s-1FILE\s0\fR. If \fI\s-1FILE\s0\fR is \fI\-\fR \f(CW\*(C`wol\*(C'\fR reads from
stdin.
.IP "\fB\-\fR" 4
.IX Item "-"
Read hardware addresses, \s-1IP\s0 addresses, optional ports and optional SecureON
password from stdin. So you can extract \fIMAC-ADDRESSES\fR from any source
and pipe the data into \f(CW\*(C`wol\*(C'\fR so.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Turns on verbose output.
.IP "\fB\-w \s-1NUM\s0\fR" 4
.IX Item "-w NUM"
.PD 0
.IP "\fB\-\-wait=NUM\fR" 4
.IX Item "--wait=NUM"
.PD
Waits \fI\s-1NUM\s0\fR milliseconds between Magic Packets. Also known as fuse health
pack.
.IP "\fB\-\-passwd[=PASS]\fR" 4
.IX Item "--passwd[=PASS]"
Send a magic packet with SecureON password feature. \fI\s-1PASS\s0\fR is written as
x\-x\-x\-x\-x\-x, where x is a hexadecimal number between 0 and ff which represents
one byte of the password.
.Sp
If you don't provide \fI\s-1PASS\s0\fR \f(CW\*(C`wol\*(C'\fR prompts you for a password.
.Sp
To set the password of your SecureON capable \s-1NIC\s0, you can use
\&\f(CW\*(C`ethtool\*(C'\fR (<\fBhttp://www.sourceforge.net/projects/gkernel/\fR>).
.SH "BUGS"
.IX Header "BUGS"
Please report bugs to <\fBkrennwallner@aon.at\fR>. Feel free to send bug
reports, translations, improvement suggestions and patches to this email
address.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\s-1GNU\s0 info entry for \fIwol\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
Written by Thomas Krennwallner <krennwallner@aon.at>.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2001,2002,2003,2004 Thomas Krennwallner <\fBkrennwallner@aon.at\fR>
.PP
This document may be distributed and modified under the terms of the \s-1GNU\s0
General Public License.
.PP
This manual is free software; you may redistribute it and/or modify it under
the terms of the \s-1GNU\s0 General Public License as published by the Free Software
Foundation; either version 2, or (at your option) any later version.
.PP
This manual is distributed in the hope that it will be useful, but without any
warranty; without even the implied warranty of merchantability or fitness for a
particular purpose. See the \s-1GNU\s0 General Public License for more details.
.PP
A copy of the \s-1GNU\s0 General Public License is available on the World Wide Web at
the \s-1GNU\s0 website. You can also obtain it by writing to the Free Software
Foundation, Inc., 59 Temple Place \- Suite 330, Boston, \s-1MA\s0 02111\-1307, \s-1USA\s0.


release/src/router/wol/src/wol.c
/*
 * wol - wake on lan client
 *
 * main program
 * 
 * $Id: wol.c,v 1.18 2004/04/18 11:42:11 wol Exp $
 *
 * Copyright (C) 2000,2001,2002,2003,2004 Thomas Krennwallner <krennwallner@aon.at>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 */



#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <error.h>

#include "wrappers.h"
#include "xalloc.h"
#include "wol.h"
#include "magic.h"
#include "net.h"
#include "macfile.h"
#include "getpass4.h"



/* My name is argv[0], used by error() */
char *program_name;

/* pointer to a MAC address */
static char *mac_str = NULL;

/* IP Address or hostname magic packet is addressed to */
static char *host_str = DEFAULT_IPADDR;

/* filename with mac addresses */
static char *pathname = NULL;

/* udp port */
static unsigned int port = DEFAULT_PORT;

/* SecureON password */
static char *passwd = NULL;

/* default is not to read from stdin */
static int request_stdin = 0;

/* be verbose */
static int verbose = 0;

/* send proxy packet */
static int proxy_mode = 0;

/* how long to wait between packets */
static int msecs = 0;

/* a magic packet */
static struct magic *magic = NULL;

/* socket file descriptor */
static int sockfd = -1;



static void
usage (int status)
{
  if (status)
    {
      fprintf (stderr, _("Try `%s --help' for more information.\n"), program_name);
    }
  else
    {
      fprintf (stdout, _("\
Usage: %s [OPTION] ... MAC-ADDRESS ...\n\
Wake On LAN client - wakes up magic packet compliant machines.\n\n\
    --help          display this help and exit\n\
-V, --version       output version information and exit\n\
-v, --verbose       verbose output\n\
-w, --wait=NUM      wait NUM millisecs after sending\n\
-h, --host=HOST     broadcast to this IP address or hostname\n\
-i, --ipaddr=HOST   same as --host\n\
-p, --port=NUM      broadcast to this UDP port\n\
-f, --file=FILE     read addresses from file FILE (\"-\" reads from stdin)\n\
    --passwd[=PASS] send SecureON password PASS (if no PASS is given, you\n\
                    will be prompted for the password)\n\
\n\
Each MAC-ADDRESS is written as x:x:x:x:x:x, where x is a hexadecimal number\n\
between 0 and ff which represents one byte of the address, which is in\n\
network byte order (big endian).\n"), program_name);

      fprintf (stdout, _("\n\
PASS is written as x-x-x-x-x-x, where x is a hexadecimal number between 0\n\
and ff which represents one byte of the password.\n"));

      fprintf (stdout, _("\nReport bugs to <krennwallner@aon.at>\n"));
    }

  exit (status);
}



static void
version (void)
{
  fprintf (stdout, PACKAGE " " VERSION "\n\n");
  fprintf (stdout, _("\
Copyright (C) 2000-2004 Thomas Krennwallner <krennwallner@aon.at>\n\
This is free software; see the source for copying conditions. There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\
\n"));
}



/* parse command line and set various globals */
static int
parse_args (int argc, char *argv[])
{
  int c;
  int option_index;
  int password_set = 0;
  char *options = "Vvw:h:i:p:f:s:-";
  static struct option long_options[] = 
    {
      { "help", no_argument, NULL, 'H' },
      { "version", no_argument, NULL, 'V' },
      { "verbose", no_argument, NULL, 'v' },
      { "wait", required_argument, NULL, 'w' },
      { "host", required_argument, NULL, 'h' },
      { "ipaddr", required_argument, NULL, 'i' },
      { "port", required_argument, NULL, 'p' },
      { "file", required_argument, NULL, 'f' },
      { "passwd", optional_argument, NULL, 'P' },
      /* { "proxy", required_argument, NULL, 's' }, */
      { NULL, 0, NULL, 0 }
    };


  if (argc == 1)
    {
      error (0, 0, _("Too few arguments."));
      usage (1);
    }



  for (;;)
    {
      c = getopt_long (argc, argv, options, long_options, &option_index);
      if (c == -1) break;

      switch (c)
	{
	case 'H':
	  usage (0);
	  break;

	  
	case 'V':
	  version ();
	  exit (0);
	  break;

	  
	case 'v':
	  verbose = 1;
	  break;


	case 'w':
	  if (sscanf (optarg, "%u", &msecs) != 1)
	    {
	      error (0, 0, _("Invalid time given"));
	      usage (1);
	    }
	  msecs *= 1000;
	  break;


	/* case 's': */
	/*   proxy_mode = 1; */
	case 'h':
	case 'i':
	  host_str = optarg;
	  break;


	case 'p':
	  if ((sscanf (optarg, "%5u", &port) != 1) ||
	      port > 65535 || port == 0)
	    {
	      error (0, 0, _("Invalid port given"));
	      usage (1);
	    }
	  break; 


	case 'f':
	  pathname = optarg;
	  break;


	case 'P':
	  if (optarg == NULL)
	    {
	      size_t n;
	      
	      if (password_set)
		break;
	      
	      if (getpass4 (_("Password"), &passwd, &n, stdin) == -1)
		{
		  error (1, 0, "getpass4 failed");
		}
	      password_set = 1;
	    }
	  else
	    {
	      passwd = optarg;
	    }
	  break;


	case '?':
	  break;
	}
    }

  if ((optind == argc) && (pathname == NULL))
    {
      error (0, 0, _("You must specify at least one MAC-ADDRESS."));
      usage (1);
    }

  /* check if stdin is requested */
  if (optind < argc)
    {
      int i;

      for (i = optind; i < argc; ++i)
	{
	  if (argv[i][0] == '-' && argv[i][1] == 0)
	    {
	      request_stdin = 1;
	      break;
	    }
	}
    }
  else if (pathname != NULL)
    {
      if (!strncmp (pathname, "-", 1))
	{
	  request_stdin = 1;
	}
    }

  /* return the offset of the GNU getopt sorted parameters */
  return optind;
}



static int
assemble_and_send (struct magic *m,
		   const char *mac_str,
		   const char *host_str,
		   unsigned int portnum,
		   const char *pass_str,
		   int socketfd)
{
  int ret = magic_assemble (m, mac_str, pass_str);
	
  switch (ret)
    {
    case -1:
      error (0, errno, _("Cannot assemble magic packet for '%s'"), mac_str);
      errno = 0;
      return -1;

    case -2:
      error (0, 0, _("Invalid password given for '%s'"), mac_str);
      errno = 0;
      return -1;
    }

  if (udp_send (socketfd, host_str, portnum, m->packet, m->size))
    {
      error (0, errno, _("Cannot send magic packet for '%s' to %s:%d"),
	     mac_str, host_str, portnum);
      errno = 0;
      return -1;
    }

  fprintf (stdout, _("Waking up %s"), mac_str);
  if (verbose)
    {
      fprintf (stdout, _(" with %s:%d"), host_str, portnum);
    }
  fprintf (stdout, _("...\n"));

  if (msecs)
    {
      usleep (msecs);
    }

  return 0;
}



int 
main (int argc, char *argv[])
{
  int i;
  int ret = 0;

  /* my name is ... */
  program_name = argv[0];

#if ENABLE_NLS
  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif /* ENABLE_NLS */
	
  i = parse_args (argc, argv);

  magic = magic_create (passwd != NULL);
  if (magic == NULL)
    {
      exit (1);
    }

  if (!proxy_mode)
    {
      sockfd = udp_open ();
    }
  else
    {
      sockfd = tcp_open (host_str, port);
    }

  if (sockfd < 0)
    {
      exit (1);
    }


  /* loop through possible MAC addresses */
  if (!request_stdin)
    {
      if (!proxy_mode)
	{
	  for (; i < argc; i++)
	    {
	      ret -= assemble_and_send (magic, argv[i], host_str, port, passwd,	sockfd);
	    }
	}
      else
	{
	  /* FIXME: proxy mode */
	}
    }


  /* -f given */
  if (pathname || request_stdin)
    {
      FILE *fp;

      if (request_stdin)
	{
	  fp = stdin;
	}
      else
	{
	  fp = fopen (pathname, "r");
	  if (fp == NULL)
	    {
	      error (1, errno, "%s", pathname);
	    }
	}

      if (!proxy_mode)
	{
	  /* loop through fp */
	  for (;;)
	    {
	      if (macfile_parse (fp, &mac_str, &host_str, &port, &passwd)) break;
	      
	      if (port == 0 || port > 65535)
		{
		  port = DEFAULT_PORT;
		}

	      ret -= assemble_and_send (magic, mac_str, host_str, port, passwd,	sockfd);
	  
	      XFREE (mac_str);
	      XFREE (host_str);
	      XFREE (passwd);
	    }
	}
      else
	{
	  /* FIXME: proxy mode */
	}

      fclose (fp);
    }

  net_close (sockfd);

  magic_destroy (magic);
  
  exit (ret != 0);
}
